public/app/features/correlations/useCorrelations.ts

```diff
@@ -37,13 +36,10 @@ export interface CorrelationsData {
   totalCount: number;
 }
 
-const toEnrichedCorrelationData = ({
-  sourceUID,
-  targetUID,
-  ...correlation
-}: Correlation): CorrelationData | undefined => {
+const toEnrichedCorrelationData = ({ sourceUID, ...correlation }: Correlation): CorrelationData | undefined => {
   const sourceDatasource = getDataSourceSrv().getInstanceSettings(sourceUID);
-  const targetDatasource = targetUID ? getDataSourceSrv().getInstanceSettings(targetUID) : undefined;
+  const targetDatasource =
+    correlation.type === 'query' ? getDataSourceSrv().getInstanceSettings(correlation.targetUID) : undefined;
```

Comment:
In general instead of checking for existence of properties we need to check what type of correlation we have and TS will then know in the branches what properties exist.